macaroonbakery/codec.py

```diff
@@ -117,4 +204,96 @@ def _decode_caveat_v1(key, caveat):
         key,
         base64.b64decode(record.get('RootKey')),
         caveat,
-        bakery.BAKERY_V1)
+        bakery.BAKERY_V1,
+        macaroon.legacy_namespace()
+    )
+
+
+def _decode_caveat_v2_v3(version, key, caveat):
+    '''Decodes a version 2 or version 3 caveat.
+    '''
+    if (len(caveat) < 1 + _PUBLIC_KEY_PREFIX_LEN +
+            _KEY_LEN + Box.NONCE_SIZE + 16):
+        raise ValueError('caveat id too short')
+    original_caveat = caveat
+    caveat = caveat[1:]  # skip version (already checked)
+
+    pk_prefix = caveat[:_PUBLIC_KEY_PREFIX_LEN]
+    caveat = caveat[_PUBLIC_KEY_PREFIX_LEN:]
+    if key.public_key.encode()[:_PUBLIC_KEY_PREFIX_LEN] != pk_prefix:
+        raise ValueError('public key mismatch')
+
+    first_party_pub = caveat[:_KEY_LEN]
+    caveat = caveat[_KEY_LEN:]
+    nonce = caveat[:Box.NONCE_SIZE]
+    caveat = caveat[Box.NONCE_SIZE:]
+    fp_public_key = PublicKey(first_party_pub)
+    box = Box(key, fp_public_key)
+    data = box.decrypt(caveat, nonce)
+    root_key, condition, ns = _decode_secret_part_v2_v3(version, data)
+    return macaroon.ThirdPartyCaveatInfo(
+        condition.decode('utf-8'),
+        fp_public_key,
+        key,
+        root_key,
+        original_caveat,
+        version,
+        ns
+    )
+
+
+def _decode_secret_part_v2_v3(version, data):
+    if len(data) < 1:
+        raise ValueError('secret part too short')
+    got_version = six.byte2int(data[:1])
+    data = data[1:]
+    if version != got_version:
+        raise ValueError(
+            'unexpected secret part version, got {} want {}'.format(
+                got_version, version))
+    root_key_length, read = _decode_varint(data)
+    data = data[read:]
+    root_key = data[:root_key_length]
+    data = data[root_key_length:]
+    if version >= bakery.BAKERY_V3:
+        namespace_length, read = _decode_varint(data)
+        data = data[read:]
+        ns_data = data[:namespace_length]
+        data = data[namespace_length:]
+        ns = namespace.deserialize_namespace(ns_data)
+    else:
+        ns = macaroon.legacy_namespace()
+    return root_key, data, ns
+
+
+def _encode_varint(i, data):
+    '''encodes integer into variable-length format into data.'''
+    while True:
+        this_byte = i & 127
+        i >>= 7
+        if i == 0:
+            data.append(this_byte)
+            break
+        data.append(this_byte | 128)
+        i -= 1
+
+
+def _decode_varint(data):
+    '''Decode a variable -length integer.
+
+    Reads a sequence of unsigned integer byte and decodes them into an integer
+    in variable-length format and returns it and the length read.
+    '''
+    i = 0
+    shift = 1
+    length = 0
+    while True:
```

Comment:
This all seems a bit more complex than it needs to be.
How about something like:

    def _decode_varint(data):
        n = 0
        shift = 0
        for b in data:
            n |= (b & 0x7f) << shift
            if (b & 0x80) == 0:
                break
            shift += 7
        return n

?